### ISSUE TITLE
Shared-QK transformer for the transformer (nn.activation.MultiheadAttention) module in PyTorch?

### ISSUE THREAD

**Issue by rudvlf0413** [pytorch/pytorch#35458]
## 🚀 Feature

We can reduce the number of parameters by sharing the linear projection layers for queries and keys in the nn.Transformer and nn.activation.MultiheadAttention modules.

I think we edit the init and forward function in nn.activation.MultiheadAttention.
 
https://github.com/pytorch/pytorch/blob/be0cdf5d157a84dd28bd95ca8f5f82306c0340f5/torch/nn/modules/activation.py#L757

https://github.com/pytorch/pytorch/blob/be0cdf5d157a84dd28bd95ca8f5f82306c0340f5/torch/nn/modules/activation.py#L853

## Motivation

The recent paper, Reformer, proposed the shared-QK Transformer that shares the linear projection layer for keys and queries. It reduces the number of parameters and memory space while there is very little performance decrease.
https://arxiv.org/abs/2001.04451

I think it is a very simple but efficient technique. If PyTorch provides this technique, many researchers can efficiently manage resources such as GPU memory.

**Comment by rudvlf0413** [comment]
However, the dot-product of a query vector with itself in the shared-QK transformer will always be greater than the dot-product of a query vector with a vector at another position.
Therefore, the paper modify the masking to prevent a token from attending to itself, except in situations where a token has no other valid attention targets (e.g. first token in a sequence).

**Comment by ngimel** [comment]
cc @zhangguanheng66

**Comment by rudvlf0413** [comment]
<img width="589" alt="image" src="https://user-images.githubusercontent.com/6302455/77738645-8db0eb80-7053-11ea-9e88-197c8602fdb0.png">

As you can see the figure, sharing linear projection layer for keys and queries does not decrease the performance (or negligible).

**Comment by rudvlf0413** [comment]
I think the chunking technique in Reformer is also easily appliable into PyTorch to reduce the memory usage for the feed-forward networks.

Authors said that "computational in feed-forward layers are completely independent across positions in a sequence, so the computation can be split into c chunks".

The technique is already implemented in PyTorch:
https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L621

**Comment by gchanan** [comment]
@zhangguanheng66 you should be making a call on this.

**Comment by zhangguanheng66** [comment]
I think this can be done with the new MHA building blocks by passing a single `nn.Linear` to the init of `MultiheadAttentionContainer`

**Comment by Kyeongpil** [comment]
@zhangguanheng66 Sorry for the late response.
I think it is hard to implement Shared-QK attention using the nn.MultiheadAttention module.
Because shared-QK attention sets k_j = \frac{q_j}{||q_j||} and prevents tokens from attending to themselves.
However, all query and key vectors are calculated using F.multi_head_attention_forward in nn.MultiheadAttention.
